### DISCUSSION TITLE
When does the pip package get updated?

### DISCUSSION THREAD

**Discussion by HolySoap** [sezanzeb/input-remapper#113] [Q&A]
The newest version is still 0.8.0 at the time I'm writing this.

```
$:~> pip install --user --upgrade key-mapper
Requirement already up-to-date: key-mapper in /usr/local/lib/python3.8/site-packages (0.8.0)
Requirement already satisfied, skipping upgrade: evdev in /usr/lib64/python3.8/site-packages (from key-mapper) (1.4.0)
Requirement already satisfied, skipping upgrade: pydbus in ./.local/lib/python3.8/site-packages (from key-mapper) (0.6.0)
Requirement already satisfied, skipping upgrade: setuptools in /usr/lib/python3.8/site-packages (from key-mapper) (44.1.1)
$:~> pip list --outdated
Package             Version  Latest       Type
------------------- -------- ------------ -----
argcomplete         1.12.2   1.12.3       wheel
click               7.1.2    8.0.1        wheel
createrepo-c        0.16.0   0.17.1.post1 sdist
cryptography        3.3.2    3.4.7        wheel
decorator           4.4.2    5.0.9        wheel
importlib-metadata  3.7.2    4.0.1        wheel
importlib-resources 3.3.0    5.1.4        wheel
iniconfig           0.0.0    1.1.1        wheel
more-itertools      8.7.0    8.8.0        wheel
ordered-set         3.1.1    4.0.2        sdist
pip                 20.2.4   21.1.1       wheel
pipx                0.14.0.0 0.16.2.1     wheel
setuptools          44.1.1   57.0.0       wheel
setuptools-scm      4.1.2    6.0.1        wheel
slip                0.6.5    20191113     sdist
userpath            1.4.1    1.5.0        wheel
virtualenv          20.2.2   20.4.6       wheel
```

If its something on my end I have to wait till openSUSE updates its packages.

**Comment by sezanzeb** [comment]
`pip install --upgrade git+https://github.com/sezanzeb/key-mapper.git`

key-mapper is not in pypi

not sure if --user installtion works btw, because files are copied into directories owned by root

**Reply by HolySoap** [reply]
Then how do I update or should I stick with it until a flatpak arrives at Flathub?

**Reply by sezanzeb** [reply]
> Then how do I update

With the command that I shared above (with sudo, forgot to add that), which can also be seen in the readme https://github.com/sezanzeb/key-mapper#pip

> until a flatpak arrives at Flathub

I don't think the flatpak package will arrive:

> If you don't have a particular privilege (for example access to /dev/uinput) before entering the Flatpak sandbox, then you still won't have that privilege inside the sandbox. If you did, that would be a security vulnerability.

(https://github.com/flatpak/flatpak/issues/4137)

**Reply by HolySoap** [reply]
Now I get errors...

```
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python3.8 /usr/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py build_wheel /tmp/tmpikt18021
       cwd: /tmp/pip-install-4fnswq3d/pycairo
  Complete output (15 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-3.8
  creating build/lib.linux-x86_64-3.8/cairo
  copying cairo/__init__.py -> build/lib.linux-x86_64-3.8/cairo
  copying cairo/__init__.pyi -> build/lib.linux-x86_64-3.8/cairo
  copying cairo/py.typed -> build/lib.linux-x86_64-3.8/cairo
  running build_ext
  Package cairo was not found in the pkg-config search path.
  Perhaps you should add the directory containing `cairo.pc'
  to the PKG_CONFIG_PATH environment variable
  Package 'cairo', required by 'virtual:world', not found
  Command '['pkg-config', '--print-errors', '--exists', 'cairo >= 1.15.10']' returned non-zero exit status 1.
  ----------------------------------------
  ERROR: Failed building wheel for pycairo
Failed to build pycairo
ERROR: Could not build wheels for pycairo which use PEP 517 and cannot be installed directly
```

There are some rules for Flatpak's to give them as much security as possible, but you are still free to use as necessary.
If you look at the [Permission Requirements](https://github.com/flathub/flathub/wiki/App-Requirements#filesystem-access) you can see that you can use file paths and device access as you wish if its necessary for the program to function correctly.
As far as I can see there is nothing to stop you from [submitting](https://github.com/flathub/flathub/wiki/App-Submission#how-to-submit-an-app) a pull request.

**Reply by sezanzeb** [reply]
If you can get a flatpak package to work feel free to share how you did that

maybe any of the following fixes the issue with cairo: https://stackoverflow.com/a/54448086, https://github.com/Automattic/node-canvas/issues/958#issuecomment-427618140, https://github.com/pygobject/pycairo/issues/89#issuecomment-600889038

**Reply by HolySoap** [reply]
I know nothing about programming, this python terminal stuff is already too much for me.

I mean the program works and I can manipulate my foot pedal how I want, that's enough for me until an easy to manage method pops up, be it as a flatpak or RPM (and actually gets in the openSUSE repos).

But anyway, thanks for the help. :)

**Reply by sezanzeb** [reply]
I unfortunately don't know how to fix that issue in openSUSE. Are you able to install the following via your package manager: `python-devel`, `gobject-introspection-devel`, `cairo-gobject-devel` somehow?

@Blaimi is working on an RPM package: https://github.com/sezanzeb/key-mapper/issues/83, don't know what the state of it is though

**Reply by HolySoap** [reply]
`cairo-gobject-devel` is not in the repos or its named differently. I'm really the wrong person for troubleshooting issues, it's probably a missing package or a package that's handled by the repo and not python.

But as I said, I can wait until something easier to manage pops up. Your awesome program is still very young and such things take time, especially when you have to learn something new to fix very specific issues.
